### PR TITLE
Add localizations for "Preferences"

### PR DESCRIPTION
**Pull request by SamusAranX**
These localizations were taken from macOS system apps and should stay accurate until the definition of the word "Preferences" changes. As far as I can tell, every supported language is covered, but in case the system is using an exotic combination of region and language codes, there's a fallback to English.

I put the ellipsized variants in their own dictionary because some of them use a localized ellipsis character. It was easier to do this than to manually construct the ellipsized strings.

### CODE REVIEW

**Comment by sindresorhus** [comment]
You missed this from the original issue:

> I would prefer a solution that doesn't force us to manually add localizations.

**Comment by SamusAranX** [comment]
macOS has no mechanism for this. Unless you take system app bundles apart to look for the system's localized strings during runtime, there's no way to do this.

**Comment by DivineDominion** [comment]
@SamusAranX Is there a way to automate how you obtained this? E.g. write a script that performs the `grep` extraction? We could add this as a build tool and then localization updates would be noticed right away -- not that I expect languages to change that much, but you never know, right :)

**Comment by SamusAranX** [comment]
I used this: https://github.com/hendriks73/extract-localization to extract various system apps' strings.
I then manually went through the resulting key/value files to find the German translation of "Preferences" and used the corresponding key with a small script to extract all other translations and put them into a neat array.
One could probably automate this, but it would be non-trivial.

File: Sources/Preferences/util.swift [hunk 1]
```diff
@@ -1,10 +1,36 @@
...
+	private static let preferencesEllipsizedLocalized = ["ar": "تفضيلات…", "ca": "Preferències…", "cs": "Předvolby…", "da": "Indstillinger…", "de": "Einstellungen…", "el": "Προτιμήσεις…", "en": "Preferences…", "en_AU": "Preferences…", "en_GB": "Preferences…", "es": "Preferencias…", "es_419": "Preferencias…", "fi": "Asetukset…", "fr": "Préférences…", "fr_CA": "Préférences…", "he": "העדפות…", "hi": "प्राथमिकता…", "hr": "Postavke…", "hu": "Beállítások…", "id": "Preferensi…", "it": "Preferenze…", "ja": "環境設定…", "ko": "환경설정...", "ms": "Keutamaan…", "nl": "Voorkeuren…", "no": "Valg…", "pl": "Preferencje…", "pt": "Preferências…", "pt_PT": "Preferências…", "ro": "Preferințe…", "ru": "Настройки…", "sk": "Nastavenia…", "sv": "Inställningar…", "th": "การตั้งค่า…", "tr": "Tercihler…", "uk": "Параметри…", "vi": "Tùy chọn…", "zh_CN": "偏好设置…", "zh_HK": "偏好設定⋯", "zh_TW": "偏好設定⋯"]
```
Comment (sindresorhus) [inline Sources/Preferences/util.swift]: Can you put each entry on separate lines?

File: Sources/Preferences/util.swift [hunk 2]
```diff
@@ -1,10 +1,36 @@
 import Cocoa
 
-struct System {
-	/// Get a system localized string
-	/// Use https://itunes.apple.com/no/app/system-strings/id570467776 to find strings
-	static func localizedString(forKey key: String) -> String {
-		return Bundle(for: NSApplication.self).localizedString(forKey: key, value: nil, table: nil)
+class System {
```
Comment (sindresorhus) [inline Sources/Preferences/util.swift]: Why did you change this to a class?

File: Sources/Preferences/util.swift [hunk 3]
```diff
@@ -1,10 +1,36 @@
...
+	private static let preferencesEllipsizedLocalized = ["ar": "تفضيلات…", "ca": "Preferències…", "cs": "Předvolby…", "da": "Indstillinger…", "de": "Einstellungen…", "el": "Προτιμήσεις…", "en": "Preferences…", "en_AU": "Preferences…", "en_GB": "Preferences…", "es": "Preferencias…", "es_419": "Preferencias…", "fi": "Asetukset…", "fr": "Préférences…", "fr_CA": "Préférences…", "he": "העדפות…", "hi": "प्राथमिकता…", "hr": "Postavke…", "hu": "Beállítások…", "id": "Preferensi…", "it": "Preferenze…", "ja": "環境設定…", "ko": "환경설정...", "ms": "Keutamaan…", "nl": "Voorkeuren…", "no": "Valg…", "pl": "Preferencje…", "pt": "Preferências…", "pt_PT": "Preferências…", "ro": "Preferințe…", "ru": "Настройки…", "sk": "Nastavenia…", "sv": "Inställningar…", "th": "การตั้งค่า…", "tr": "Tercihler…", "uk": "Параметри…", "vi": "Tùy chọn…", "zh_CN": "偏好设置…", "zh_HK": "偏好設定⋯", "zh_TW": "偏好設定⋯"]
+
+	static func localizedPreferences(ellipsized: Bool = false) -> String {
```
Comment (sindresorhus) [inline Sources/Preferences/util.swift]: This method is too specific. It should be possible to add additional localizations in the future.

I'm thinking something like `static func get(string: Localization.Identifier)`, and used like `Localization.get(.preferences)`.

File: Sources/Preferences/util.swift (hunk 2 above)
Comment (sindresorhus) [reply Sources/Preferences/util.swift]: I think it would be more proper to call it `Localization` now.

File: Sources/Preferences/util.swift [hunk 4]
```diff
@@ -1,10 +1,36 @@
...
+				if let localizedString = System.preferencesLocalized[identifier] {
+					return localizedString
+				}
+				if let localizedString = System.preferencesLocalized[languageCode] {
+					return localizedString
+				}
+			}
+		}
+
+		// fallback to English locale
```
Comment (sindresorhus) [inline Sources/Preferences/util.swift]:
```suggestion
		// Fall back to English locale
```

**Comment by sindresorhus** [comment]
Can you ensure the localization resolution logic follows the logic used when using `Localizable.strings`?

File: Sources/Preferences/util.swift (hunk 1 above)
Comment (SamusAranX) [reply Sources/Preferences/util.swift]: done

File: Sources/Preferences/util.swift (hunk 3 above)
Comment (SamusAranX) [reply Sources/Preferences/util.swift]: I implemented the identifiers as an enum to ensure that people don't pass invalid values to the `.get(identifier)` method.

**Comment by SamusAranX** [comment]
I implemented most of the suggested changes, but I'm not sure what you mean regarding the localization resolution logic.
The `Localization.get()` method's designed to be pretty robust, returning correctly localized strings no matter what locale your system uses.
Unless there's plans to create an exhaustive localization dictionary, I think the "identifier: string" model is the best one for this task.

File: Sources/Preferences/Localization.swift [hunk 5]
```diff
@@ -0,0 +1,114 @@
+import Foundation
+
+public enum LocalizationIdentifier {
```
Comment (sindresorhus) [inline Sources/Preferences/Localization.swift]: Move this into the `Localization` struct as `Identifier`.

File: Sources/Preferences/Localization.swift [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	case preferences
+	case preferencesEllipsized
+}
+
+struct Localization {
+	private static let localizedStrings: [LocalizationIdentifier: [String: String]] = [
```
Comment (sindresorhus) [inline Sources/Preferences/Localization.swift]:
```suggestion
	private static let localizedStrings: [Identifier: [String: String]] = [
```

File: Sources/Preferences/Localization.swift [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+			"tr": "Tercihler…",
+			"uk": "Параметри…",
+			"vi": "Tùy chọn…",
+			"zh_CN": "偏好设置…",
+			"zh_HK": "偏好設定⋯",
+			"zh_TW": "偏好設定⋯"
+		]
+	]
+
+	static func get(identifier: LocalizationIdentifier) -> String {
```
Comment (sindresorhus) [inline Sources/Preferences/Localization.swift]:
```suggestion
	static func get(identifier: Identifier) -> String {
```

And would be useful with a short doc comment here.

File: Sources/Preferences/Localization.swift [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+		let locale = Locale.current
+		if let languageCode = locale.languageCode, let regionCode = locale.regionCode, let localizedDict = Localization.localizedStrings[identifier] {
```
Comment (sindresorhus) [inline Sources/Preferences/Localization.swift]: I would use a `guard` here instead.

File: Sources/Preferences/Localization.swift [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+			if let localizedString = localizedDict[languageCode] {
+				return localizedString
+			}
+
+			// Fall back to English locale, which always exists
+			return localizedDict["en"]!
+		}
+
+		// Fall back that shouldn't ever happen
+		return "PREFERENCES"
```
Comment (sindresorhus) [inline Sources/Preferences/Localization.swift]: This is too specific to the current use-case. We will add other strings in the future.

File: Sources/Preferences/Localization.swift (hunk 8 above)
Comment (SamusAranX) [reply Sources/Preferences/Localization.swift]: Makes sense.

File: Sources/Preferences/Localization.swift (hunk 9 above)
Comment (SamusAranX) [reply Sources/Preferences/Localization.swift]: I'm torn between just force-unwrapping the `Localization.localizedStrings[identifier]` part and the current solution.
On the one hand, this code is completely out of the hands of the end users, so we could get away with force-unwrapping here, which would eliminate this last fallback return. On the other hand… well, there's the force-unwrapping 😬

Comment (sindresorhus) [reply Sources/Preferences/Localization.swift]: I would just force-unwrap then, but add a comment explaining why force-unwrapping is ok there.

**Comment by SamusAranX** [comment]
I'm not sure if Github sends out notifications for commits, so here's a comment for good measure.

I restructured the code a bit and added a documentation comment to the `get()` method.

**Comment by sindresorhus** [comment]
Looks good now. Thank you! :)

**Comment by sindresorhus** [comment]
@SamusAranX I noticed a problem after merging this. If the user sets their language as `zh`, but their region as `US` (United States), the identifier will be `zh-US` and not match anything. We need to handle that case too.

From some reading, it also seems like we should respect https://developer.apple.com/documentation/foundation/locale/2293155-preferredlanguages

Some more reading here: https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html

**Comment by SamusAranX** [comment]
That's an interesting problem that I admittedly didn't think of.

The question now becomes:
If the user has their system set to a `language-region-script` combination that doesn't match, for example `zh-US-Hans`, which language should ultimately be returned? English (US) or Chinese? If Chinese, which Chinese localization from the set of `["zh-CN", "zh-HK", "zh-TW"]` should be chosen?

I just checked and there's no first-party app containing just a `zh.lproj" localization, so I don't know which flavor of `zh` Apple defaults to.

I could build a generic mechanism that just falls back to the first specified language in alphabetical order if the specific region code combination can't be found, but that probably wouldn't be that users would expect.
